v2/block_reader.go

```diff
@@ -0,0 +1,142 @@
+// See NewBlockReader
+type BlockReader struct {
+	// The detected version of the CAR payload.
+	Version uint64
+	// The roots of the CAR payload. May be empty.
+	Roots []cid.Cid
+
+	// Used internally only, by BlockReader.Next during iteration over blocks.
+	r     io.Reader
+	ropts ReadOptions
+}
+
+// NewBlockReader instantiates a new BlockReader facilitating iteration over blocks in CARv1 or
+// CARv2 payload. Upon instantiation, the version is automatically detected and exposed via
+// BlockReader.Version. The root CIDs of the CAR payload are exposed via BlockReader.Roots
+//
+// See BlockReader.Next
+func NewBlockReader(r io.Reader, opts ...ReadOption) (*BlockReader, error) {
+	// Read CARv1 header or CARv2 pragma.
+	// Both are a valid CARv1 header, therefore are read as such.
+	pragmaOrV1Header, err := carv1.ReadHeader(r)
+	if err != nil {
+		return nil, err
+	}
+
+	// Populate the block reader version.
+	br := &BlockReader{
+		Version: pragmaOrV1Header.Version,
+	}
+
+	// Populate read options
+	for _, o := range opts {
+		o(&br.ropts)
+	}
+
+	// Expect either version 1 or 2.
+	switch br.Version {
+	case 1:
+		// If version is 1, r represents a CARv1.
+		// Simply populate br.Roots and br.r without modifying r.
+		br.Roots = pragmaOrV1Header.Roots
+		br.r = r
+	case 2:
+		// If the version is 2:
+		//  1. Read CARv2 specific header to locate the inner CARv1 data payload offset and size.
+		//  2. Skip to the beginning of the inner CARv1 data payload.
+		//  3. Re-initialize r as a LimitReader, limited to the size of the inner CARv1 payload.
+		//  4. Read the header of inner CARv1 data payload via r to populate br.Roots.
+
+		// Read CARv2-specific header.
+		v2h := Header{}
+		if _, err := v2h.ReadFrom(r); err != nil {
+			return nil, err
+		}
+		// Assert the data payload offset validity.
+		// It must be at least 51 (<CARv2Pragma> + <CARv2Header>).
+		dataOffset := int64(v2h.DataOffset)
+		if dataOffset < PragmaSize+HeaderSize {
+			return nil, fmt.Errorf("invalid data payload offset: %v", dataOffset)
+		}
+		// Assert the data size validity.
+		// It must be larger than zero.
+		// Technically, it should be at least 11 bytes (i.e. a valid CARv1 header with no roots) but
+		// we let further parsing of the header to signal invalid data payload header.
+		dataSize := int64(v2h.DataSize)
+		if dataSize <= 0 {
+			return nil, fmt.Errorf("invalid data payload size: %v", dataSize)
+		}
+
+		// Skip to the beginning of inner CARv1 data payload.
+		// Note, at this point the pragma and CARv1 header have been read.
+		// An io.ReadSeeker is opportunistically constructed from the given io.Reader r.
+		// The constructor does not take an initial offset, so we use Seek in io.SeekCurrent to
+		// fast forward to the beginning of data payload by subtracting pragma and header size from
+		// dataOffset.
+		rs := internalio.ToByteReadSeeker(r)
+		if _, err := rs.Seek(dataOffset-PragmaSize-HeaderSize, io.SeekCurrent); err != nil {
```

Comment:
I think the math here might be easier if you use SeekStart?

Reply:
I tried to explain why `SeekCurrent` is used in the comments above. TLDR; using `SeekStart` _could_ advance too far if `r` does not implement `io.ReadSeeker`. The opportunistic conversion from `io.Reader` to `io.ReadSeeker` does not take an initial offset, because we don't know if we can take an initial offset.

Reply:
Ah, I see. Makes sense to me.